src/main/java/org/junit/rules/TemporaryFolder.java

```diff
@@ -216,10 +219,23 @@ public File newFolder() throws IOException {
     }
 
     private File createTemporaryFolderIn(File parentFolder) throws IOException {
-        File createdFolder = File.createTempFile("junit", "", parentFolder);
-        createdFolder.delete();
-        createdFolder.mkdir();
-        return createdFolder;
+        File createdFolder = null;
+        for (int i = 0; i < TEMP_DIR_ATTEMPTS; ++i) {
+            // Use createTempFile to get a suitable folder name.
+            final String suffix = ".tmp";
+            final File tmpFile = File.createTempFile(TMP_PREFIX, suffix, parentFolder);
+            tmpFile.delete();
```

Comment:
You pointed out that the specification of File.createTempFile says "Neither this method nor any of its variants will return the same abstract pathname again in the current invocation of the virtual machine". I checked JDK6 (it returns random filenames: http://hg.openjdk.java.net/jdk6/jdk6/jdk/file/b72e7b89dda9/src/share/classes/java/io/File.java#l1795), and GNU Classpath (they use a mix of a timestamp and a global counter to prevent returning the same name in the lifetime of the VM). Oracle's JDK5 is not public, but I doubt there were significant changes.

I'll remove the finals.